libvast/include/vast/plugin.hpp

```diff
@@ -398,6 +398,44 @@ class rest_endpoint_plugin : public virtual plugin {
     = 0;
 };
 
+// -- loader plugin -----------------------------------------------------
+
+using input_loader = std::function<auto()->generator<chunk_ptr>>;
+using input_parser = std::function<auto(chunk_ptr)->generator<table_slice>>;
+using options = record;
+
+/// A loader plugin transfers input data into a stream of chunks.
+/// @relates plugin
+class loader_plugin : public virtual plugin {
+public:
+  /// Returns the loader.
+  [[nodiscard]] virtual auto make_loader(options, const bool_operator*) const
+    -> caf::expected<input_loader>
+    = 0;
+
+  /// Returns the default parser for this loader.
+  [[nodiscard]] virtual auto
+  make_default_parser(options, const bool_operator*) const
+    -> caf::expected<input_parser>
+    = 0;
+};
+
+class stdin_loader_plugin : public virtual loader_plugin {
+public:
+  [[nodiscard]] auto make_loader(options, const bool_operator*) const
+    -> caf::expected<input_loader> override;
+
+  [[nodiscard]] auto make_default_parser(options, const bool_operator*) const
+    -> caf::expected<input_parser> override;
+
+  [[nodiscard]] caf::error initialize(data config) override;
+
+  [[nodiscard]] std::string name() const override;
+
+  static constexpr inline auto max_chunk_size = size_t{16384};
+  static constexpr inline auto read_timeout = std::chrono::milliseconds{5000};
```

Comment:
How long should the stdin loader timeout be? I arbitrarily chose 5 seconds, but I don't know what value would be truly sensible for our users' use cases.

Reply:
For the current readers this is configurable as `vast.import.read-timeout`, which I believe defaults to 20ms. We probably need to talk about options handling, I don't think we've considered that in the loader plugin design so far.

Reply:
It would be good to first write down exactly what the timeout does. Someone who sees `read_timeout` needs more context to make a decision of what they want. Ideally it doesn't require going to the call sites.

Reply:
The `read_timeout` is passed on to the `fdinbuf` timeout value. If the `fdinbuf` file descriptor polling does not pass within that timeframe, there is a timeout. In that case it would make sense to add some coherence by unifying this setting with `vast.import.read-timeout`.

I wonder whether the operator control plane or the pipeline executor would be able to pass on option values somehow.

Reply:
If `plugin` already contains a subset of the options, maybe it can also be expanded to also include global VAST options for easy access.

Reply:
I agree generally, but that should probably not be in this PR but rather in a separate one. 

Right now, there's a member function of the `plugin` base class that gets just the relevant subsection:

```cpp
/// Initializes a plugin with its respective entries from the YAML config
/// file, i.e., `plugin.<NAME>`.
/// @param config The relevant subsection of the configuration.
[[nodiscard]] virtual caf::error initialize(data config) = 0;
```

We could consider extending this to get all options, and require the plugins themselves to access the subsection relevant to them. Since this is a bigger change that affects everyone, I would recommend starting a separate discussion in Slack and if agreed upon creating a separate PR.

The alternative would be to use a workaround like `caf::get_or(content(ctrl.self()->config()), "vast.read-timeout", 20ms)`, but that mostly works by coincidence and not be design.

Reply:
I created a draft PR that implements the plugin option extension: #2949